src/Microsoft.SqlTools.ServiceLayer/Scripting/ScriptingObjectMatchProcessor.cs

```diff
@@ -0,0 +1,111 @@
+﻿//
+// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+//
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using Microsoft.SqlTools.ServiceLayer.Scripting.Contracts;
+
+namespace Microsoft.SqlTools.ServiceLayer.Scripting
+{
+    public static class ScriptingObjectMatchProcessor
+    {
+        public static IEnumerable<ScriptingObject> Match(
+            ScriptingObject includeCriteria,
+            ScriptingObject excludeCriteria,
+            IEnumerable<ScriptingObject> candidates)
+        {
+            return Match(
+                includeCriteria == null ? new ScriptingObject[0] : new[] { includeCriteria },
+                excludeCriteria == null ? new ScriptingObject[0] : new[] { excludeCriteria },
+                candidates);
+        }
+
+        public static IEnumerable<ScriptingObject> Match(
+            IEnumerable<ScriptingObject> includeCriteria,
+            IEnumerable<ScriptingObject> excludeCriteria,
+            IEnumerable<ScriptingObject> candidates)
+        {
+            if (candidates == null)
+            {
+                throw new ArgumentNullException("candidates");
+            }
+
+            IEnumerable<ScriptingObject> matchedObjects = new List<ScriptingObject>();
+
+            if (includeCriteria != null && includeCriteria.Count() > 0)
+            {
+                foreach (ScriptingObject scriptingObjectCriteria in includeCriteria)
+                {
+                    IEnumerable<ScriptingObject> matches = MatchCriteria(scriptingObjectCriteria, candidates);
+                    matchedObjects = matchedObjects.Union(matches);
```

Comment:
Actually, I see why this is done the way it is. I think the right thing is to check .Any(), which only accesses the 1st element.

---
In reply to: [111654045](https://github.com/Microsoft/sqltoolsservice/pull/316#discussion_r111654045) [](ancestors = 111654045)